bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleRegistryImpl.java

```diff
@@ -91,87 +91,166 @@
+     *
+     * @param templateRegistry templateRegistry service.
+     */
+    protected void unsetTemplateRegistry(TemplateRegistry templateRegistry) {
+        this.templateRegistry = null;
+    }
+
+    protected void setStorageService(StorageService storageService) {
+        // this.storageService = storageService;
+        setDisabledRuleStorage(
+                storageService.<Boolean> getStorage("automation_rules_disabled", this.getClass().getClassLoader()));
+        setManagedProvider(new ManagedRuleProvider(storageService));
     }
 
     /**
-     * This method is used to register all {@link Rule}s provided via the {@link RuleProvider}, into the
-     * {@link RuleEngine}.
+     * Unbind the {@link StorageService} - called from DS.
      *
-     * @param provider a provider of {@link Rule}s.
-     * @throws RuntimeException
-     *             when passed module has a required configuration property and it is not specified in rule definition
-     *             nor
-     *             in the module's module type definition.
-     * @throws IllegalArgumentException
-     *             when a module id contains dot or when the rule with the same UID already exists.
+     * @param storageService
+     */
+    protected void unsetStorageService(StorageService storageService) {
+        // this.storageService = null;
+        removeManagedProvider(managedProvider);
+    }
+
+    /**
+     *
+     * @param templateProvider
      */
+    protected void addTemplateProvider(TemplateProvider templateProvider) {
+        templateUpdated(templateProvider.getTemplates(null));
+    }
+
+    protected void updateTemplateProvider(TemplateProvider templateProvider) {
+        templateUpdated(templateProvider.getTemplates(null));
+    }
+
+    @Override
+    protected void setEventPublisher(EventPublisher eventPublisher) {
+        super.setEventPublisher(eventPublisher);
+    }
+
+    @Override
+    protected void unsetEventPublisher(EventPublisher eventPublisher) {
+        super.unsetEventPublisher(eventPublisher);
+    }
+
     @Override
     protected void addProvider(Provider<Rule> provider) {
-        logger.info("Rule provider: {} is added.", provider);
+        LOGGER.info("Rule provider: {} is added.", provider);
         super.addProvider(provider);
+        if (provider instanceof ManagedRuleProvider) {
+            ruleRegistryReg = bundleContext.registerService(RuleRegistry.class.getName(), this, null);
+        }
     }
 
     @Override
-    protected void setManagedProvider(ManagedProvider<Rule, String> provider) {
-        super.setManagedProvider(provider);
-        logger.info("Rule Managed Provider: {} is added.", provider);
+    protected void removeProvider(Provider<Rule> provider) {
+        LOGGER.info("Rule provider: {} is removed.", provider);
+        if (provider instanceof ManagedRuleProvider) {
+            if (ruleRegistryReg != null) {
+                ruleRegistryReg.unregister();
```

Comment:
Old implementation was dependent on ManagedRuleProvider (Storage service is available)
I saw that AbstractRegistry handles situtation on **add/updated/remove** operations when ManagedProvider has **null** value (e.g. throws runtime exception).


Do we need this dependency here?